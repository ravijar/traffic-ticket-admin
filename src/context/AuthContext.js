import { createContext, useState, useEffect } from "react";
import jwt_decode from "jwt-decode";
import { useNavigate } from "react-router-dom";
import { API_URL } from "../constants/urls";

const AuthContext = createContext();

export default AuthContext;

export const AuthProvider = ({ children }) => {
  let [authTokens, setAuthTokens] = useState(() =>
    localStorage.getItem("authTokens")
      ? JSON.parse(localStorage.getItem("authTokens"))
      : null
  );
  let [user, setUser] = useState(() =>
    localStorage.getItem("authTokens")
      ? jwt_decode(localStorage.getItem("authTokens"))
      : null
  );
  let [loading, setLoading] = useState(false);

  let navigate = useNavigate();

  // user sign in
  let loginUser = async (userName, password, setLoading) => {
    setLoading(true);
    let response = await fetch(`${API_URL}/api/token/`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        username: userName,
        password: password,
      }),
    });

    let data = await response.json();
    setLoading(false);
    if (response.status === 200) {
      // checking for the user role
      if (jwt_decode(JSON.stringify(data)).role !== "admin") {
        alert("Access Denied!");
      } else {
        setAuthTokens(data);
        setUser(jwt_decode(data.access));
        localStorage.setItem("authTokens", JSON.stringify(data));
        console.log(jwt_decode(JSON.stringify(data)));
        navigate("/");
      }
    } else if (response.status === 401) {
      alert("Invalid credentials!");
    } else {
      alert("Something went wrong!");
    }
  };

  // user sign out
  let logoutUser = () => {
    setAuthTokens(null);
    setUser(null);
    localStorage.removeItem("authTokens");
    navigate("/login");
  };

  // updating refresh token
  let updateToken = async () => {
    let response = await fetch(`${API_URL}/api/token/refresh/`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        refresh: authTokens?.refresh,
      }),
    });

    let data = await response.json();
    if (response.status === 200) {
      setAuthTokens(data);
      setUser(jwt_decode(data.access));
      localStorage.setItem("authTokens", JSON.stringify(data));
    } else {
      logoutUser();
    }
    if (loading) {
      setLoading(false);
    }
  };

  let contextData = {
    user: user,
    authTokens: authTokens,
    loginUser: loginUser,
    logoutUser: logoutUser,
  };

  useEffect(() => {
    if (loading) {
      updateToken();
    }
    let time = 1000 * 60 * 4;
    let interval = setInterval(() => {
      if (authTokens) {
        updateToken();
      }
    }, time);
    return () => clearInterval(interval);
  }, [authTokens, loading]);

  return (
    <AuthContext.Provider value={contextData}>
      {loading ? null : children}
    </AuthContext.Provider>
  );
};
