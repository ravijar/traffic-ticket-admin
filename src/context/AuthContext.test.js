import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { AuthProvider } from './AuthContext';
import AuthContext from './AuthContext';

// Mock the `useNavigate` function
jest.mock('react-router-dom', () => {
  return {
    useNavigate: () => (url) => url,
  };
});

describe('AuthContext', () => {
  it('provides initial context with no user', () => {
    const { container } = render(
      <AuthProvider>
        <AuthContext.Consumer>
          {(value) => {
            return (
              <>
                <div>Username: {value.user?.username}</div>
                <div>Token: {value.authTokens?.token}</div>
              </>
            );
          }}
        </AuthContext.Consumer>
      </AuthProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it('logs out a user and clears the context', () => {
    const { container } = render(
      <AuthProvider>
        <AuthContext.Consumer>
          {(value) => {
            return (
              <>
                <button onClick={() => value.logoutUser()}>Logout</button>
                <div>Username: {value.user?.username}</div>
                <div>Token: {value.authTokens?.token}</div>
              </>
            );
          }}
        </AuthContext.Consumer>
      </AuthProvider>
    );

    const logoutButton = screen.getByText('Logout');
    fireEvent.click(logoutButton);

    expect(container).toMatchSnapshot();
  });
});
