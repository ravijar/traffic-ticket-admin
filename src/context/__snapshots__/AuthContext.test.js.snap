// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthContext logs in a user and sets the context 1`] = `
<div>
  <button>
    Login
  </button>
  <div>
    Username: 
  </div>
  <div>
    Token: 
  </div>
</div>
`;

exports[`AuthContext logs out a user and clears the context 1`] = `
<div>
  <button>
    Logout
  </button>
  <div>
    Username: 
  </div>
  <div>
    Token: 
  </div>
</div>
`;

exports[`AuthContext provides initial context with no user 1`] = `
<div>
  <div>
    Username: 
  </div>
  <div>
    Token: 
  </div>
</div>
`;
